### ISSUE TITLE
Problems receiving topic subscription events [onData, onInfo, onMeta,...]

### ISSUE THREAD

**Issue by ejbp** [tinode/tinode-js#23]
**Problem:**

Right after subscribing to a topic, tinode doesn't trigger my handlers  [onData, onInfo, onMeta,...].

**Code Example:**

```
const topic="usrIfaOwzqcliA";
const topicObj = tinode.getTopic(topic);

const getQuery = topicObj
        .startMetaQuery()
        .withLaterDesc()
        .withLaterSub()
        .withLaterData(20)
        .withLaterDel()
        .build();

let setQuery;
      if (newGroupTopicParams) {
        setQuery = newGroupTopicParams;
      } else {
        const isOwner =
          topicObj && topicObj.acs && topicObj.acs.isOwner ? topicObj.acs.isOwner() : false;
        setQuery = {
          sub: {
            mode: `JRWPASD${isOwner ? "O" : ""}`
          }
        };
      }

topicObj
          .subscribe(getQuery, setQuery)
          .then(ctrl => {
            topicObj.onData = chatApi.handlers.onTopicData(ctrl.topic);
            topicObj.onInfo = chatApi.handlers.onTopicInfo(ctrl.topic);
            topicObj.onMeta = chatApi.handlers.onTopicMeta(ctrl.topic);
            topicObj.onMetaSub = chatApi.handlers.onTopicMetaSub(ctrl.topic);
            topicObj.onAllMessagesReceived = chatApi.handlers.onAllMessagesReceived(ctrl.topic);
            return { topic: ctrl.topic };
          })

```

**Why?**

note: for better reading I will address a message received in  `this._connection.onMessage` as "in" message.

After some research, I figure out that in tinode `this._connection.onMessage`  function after the in

` in: {"ctrl":{"id":"117471","topic":"usrIfaOwzqcliA","params":{"acs":{"given":31,"want":31,"mode":31}},"code":200,"text":"ok","ts":"2019-03-21T21:22:32.951Z"}}`

it triggers the

`execPromise(pkt.ctrl.id, pkt.ctrl.code, pkt.ctrl, pkt.ctrl.text);`

which consequently is expected that the code snippet below runs in first place, before any further "in" messages.  

```
  subscribe: function(getParams, setParams) {
    // If the topic is already subscribed, return resolved promise
    if (this._subscribed) {
      return Promise.resolve(this);
    }

    // Send subscribe message, handle async response.
    // If topic name is explicitly provided, use it. If no name, then it's a new group topic,
    // use "new".
    return this._tinode.subscribe(this.name || TOPIC_NEW, getParams, setParams).then((ctrl) => {
      if (ctrl.code >= 300) {
        // Do nothing ff the topic is already subscribed to.
        return ctrl;
      }
  [...]
  
  this._cachePutSelf();

 [...]

```
  
which is happening in most of the times. Problem is, when a new "in" message enters first in the process ticket than the execution of the resolve() itself, which will make this (e.g.)

`const topic = cacheGet('topic', pkt.data.topic);` 

return null since the "resolve()" was not executed yet because it's still in a future ticket to be executed and so the `this._cachePutSelf()` was not executed yet. 

For better understanding run this code on your browser console:

```
{

const log = console.log;

const start = performance.now(); 
const promise = new Promise( (resolve, reject) => {
  
	setTimeout(function() {
	  const a = performance.now()-start;
	  log("Executing promisse", a);
    
           resolve(performance.now()-start)

	  const b = performance.now()-start;
	  console.log("After calling resolve()", b);

	}, 2000);
  
 	console.log("Please wait 2s....")
  
});

promise.then((arg)=>{
  const a = performance.now()-start;
  log("Promise resolved!", arg + "|" + a);
});

}
```

Result:

<img width="632" alt="Screenshot 2019-03-21 at 21 49 22" src="https://user-images.githubusercontent.com/2060105/54787299-3cda0c00-4c23-11e9-8be8-d00ad4ed495b.png">

You can see that the log "After calling resolve()" runs first then the "Promise resolved".  The reason why, is that "resolve" is in fact a non blocking execution, instead when we call "resolve()" we are telling the browser to set the status of our Promise to resolved but the callback itself it's enqueued as a microtask to be processed later.

 You can check the Promise chrome implementation to confirm: 

[https://chromium.googlesource.com/v8/v8/+/3.29.45/src/promise.js?autodive=0/](https://chromium.googlesource.com/v8/v8/+/3.29.45/src/promise.js?autodive=0/)

A nice reading about this topic: [https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/](https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/)

**detected on device:**

I detected this when I was testing Tinode in a Tizen Smart Tv app, on the web simulator. Here is the Navigator info:

![Screenshot 2019-03-21 at 22 03 58](https://user-images.githubusercontent.com/2060105/54787944-46fd0a00-4c25-11e9-8753-00f8fa6e985a.png)

Here is the simulator version:

![Screenshot 2019-03-21 at 22 03 22](https://user-images.githubusercontent.com/2060105/54787959-554b2600-4c25-11e9-87a4-78c21c426866.png)

**How I made it work?**
 
Please note that  **this is not my suggestion of implementation.** It was a quick test to see if the problem was in fact what I've described here:

I basically added a setTimeout ( ()=>{ //code }, 0 ) after receiving "in" messages, except when it's to resolve the promise in the `this._connection.onMessage`:

```
// The main message dispatcher.
  this._connection.onMessage = (data) => {
    // Skip empty response. This happens when LP times out.
    if (!data) return;

    this._inPacketCount++;

    // Send raw message to listener
    if (this.onRawMessage) {
      this.onRawMessage(data);
    }

    if (data === '0') {
      // Server response to a network probe.
      if (this.onNetworkProbe) {
        this.onNetworkProbe();
      }
      // No processing is necessary.
      return;
    }

    let pkt = JSON.parse(data, jsonParseHelper);
    if (!pkt) {
      this.logger("in: " + data);
      this.logger("ERROR: failed to parse data");
    } else {
      this.logger("in: " + (this._trimLongStrings ? JSON.stringify(pkt, jsonLoggerHelper) : data));

      // Send complete packet to listener
      if (this.onMessage) {
        this.onMessage(pkt);
      }

      if (pkt.ctrl) {
        // Handling {ctrl} message
        if (this.onCtrlMessage) {
          this.onCtrlMessage(pkt.ctrl);
        }

        // Resolve or reject a pending promise, if any
        if (pkt.ctrl.id) {
          execPromise(pkt.ctrl.id, pkt.ctrl.code, pkt.ctrl, pkt.ctrl.text);
        }

        setTimeout(()=>{  //<---------------------------------ADDED

          if (pkt.ctrl.code == 205 && pkt.ctrl.text == 'evicted') {
            // User evicted from topic.
            const topic = cacheGet('topic', pkt.ctrl.topic);
            if (topic) {
              topic._resetSub();
            }
          } else if (pkt.ctrl.params && pkt.ctrl.params.what == 'data') {
            // All messages received: "params":{"count":11,"what":"data"},
            const topic = cacheGet('topic', pkt.ctrl.topic);
            if (topic) {
              topic._allMessagesReceived(pkt.ctrl.params.count);
            }
          }


        }, 0);


      }else {

        setTimeout(()=>{  //<---------------------------------ADDED

          if (pkt.meta) {
            // Handling a {meta} message.
    
            // Preferred API: Route meta to topic, if one is registered
            const topic = cacheGet('topic', pkt.meta.topic);
            if (topic) {
              topic._routeMeta(pkt.meta);
            }
    
            // Secondary API: callback
            if (this.onMetaMessage) {
              this.onMetaMessage(pkt.meta);
            }
          } else if (pkt.data) {
            // Handling {data} message
    
            // Preferred API: Route data to topic, if one is registered
            const topic = cacheGet('topic', pkt.data.topic);
            if (topic) {
              topic._routeData(pkt.data);
            }
    
            // Secondary API: Call callback
            if (this.onDataMessage) {
              this.onDataMessage(pkt.data);
            }
          } else if (pkt.pres) {
            // Handling {pres} message
    
            // Preferred API: Route presence to topic, if one is registered
            const topic = cacheGet('topic', pkt.pres.topic);
            if (topic) {
              topic._routePres(pkt.pres);
            }
    
            // Secondary API - callback
            if (this.onPresMessage) {
              this.onPresMessage(pkt.pres);
            }
          } else if (pkt.info) {
            // {info} message - read/received notifications and key presses
    
            // Preferred API: Route {info}} to topic, if one is registered
            const topic = cacheGet('topic', pkt.info.topic);
            if (topic) {
              topic._routeInfo(pkt.info);
            }
    
            // Secondary API - callback
            if (this.onInfoMessage) {
              this.onInfoMessage(pkt.info);
            }
          } else {
            this.logger("ERROR: Unknown packet received.");
          }







        }, 0);


      } 
      
    }
  }
```

After this change, it started to work as expected.

**Comment by or-else** [comment]
Wow! Excellent analysis, thanks! Let me think about it a bit.

**Comment by or-else** [comment]
Let me see that I understand your use case correctly. Given code like this `chatApi.handlers.onTopicData(ctrl.topic)` I assume you generate handler functions tied to topic names. So, if you have topic name defined upfront then code like this would just work (using your example), right?:
 
```js
const topic="usrIfaOwzqcliA";
const topicObj = tinode.getTopic(topic);

// ...
// set set and get queries like before
// ...

topicObj.onData = chatApi.handlers.onTopicData(topic);
topicObj.onInfo = chatApi.handlers.onTopicInfo(topic);
topicObj.onMeta = chatApi.handlers.onTopicMeta(topic);
topicObj.onMetaSub = chatApi.handlers.onTopicMetaSub(topic);
topicObj.onAllMessagesReceived = chatApi.handlers.onAllMessagesReceived(topic);

topicObj.subscribe(getQuery, setQuery);
```

The problem happens only when the topic name is not know upfront, i.e. when one creates a new group topic?

**Comment by or-else** [comment]
The change you are suggesting is actually useful even in general. I'll do some testing to make sure nothing breaks.

**Comment by ejbp** [comment]
> Let me see that I understand your use case correctly. Given code like this `chatApi.handlers.onTopicData(ctrl.topic)` I assume you generate handler functions tied to topic names. So, if you have topic name defined upfront then code like this would just work (using your example), right?:
> 
> ```js
> const topic="usrIfaOwzqcliA";
> const topicObj = tinode.getTopic(topic);
> 
> // ...
> // set set and get queries like before
> // ...
> 
> topicObj.onData = chatApi.handlers.onTopicData(topic);
> topicObj.onInfo = chatApi.handlers.onTopicInfo(topic);
> topicObj.onMeta = chatApi.handlers.onTopicMeta(topic);
> topicObj.onMetaSub = chatApi.handlers.onTopicMetaSub(topic);
> topicObj.onAllMessagesReceived = chatApi.handlers.onAllMessagesReceived(topic);
> 
> topicObj.subscribe(getQuery, setQuery);
> ```
> 
> The problem happens only when the topic name is not known upfront, i.e. when one creates a new group topic?

Exactly, and that's the reason why the handlers are only settled inside that promise. If I settle them outside, will not work in that specific use case.  Though, if I settle inside, they are not triggered due to what I've described. In fact, contrary to what I have described, Tinode finds the topic in most of the cases

`const topic = cacheGet('topic', pkt.data.topic); //it is found`

Though, my handlers are not triggered and of course, it may give problems, when you create a new group, cause the variable topic changes as you've mentioned.

**Comment by or-else** [comment]
Unless something else breaks, I'll make the change you are suggesting. 

Meanwhile you may work around the problem by changing the 
 `chatApi.handlers.onTopicData(topic)` to something like `chatApi.handlers.onTopicData(newTopicName, optionalOldTopicName)`. First register the handler before `subscribe()` with the temporart name (newXXXX) then re-register the handler under the permanent name in the promise on `then()`. For example Java SDK does it this way:
https://github.com/tinode/tindroid/blob/master/tinodesdk/src/main/java/co/tinode/tinodesdk/Topic.java#L685

**Comment by or-else** [comment]
This has been released in v0.15.14-rc1
